File: SMO/DatabaseDefrag/CS/DatabaseDefrag/DatabaseDefrag.cs
/*============================================================================
  File:    DatabaseDefrag.cs 

  Summary: Implements a sample SMO SQL Server database defragment utility in C#.

  Date:    August 29, 2005
------------------------------------------------------------------------------
  This file is part of the Microsoft SQL Server Code Samples.

  Copyright (C) Microsoft Corporation.  All rights reserved.

  This source code is intended only as a supplement to Microsoft
  Development Tools and/or on-line documentation.  See these other
  materials for detailed information regarding Microsoft code samples.

  THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY
  KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
  IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
  PARTICULAR PURPOSE.
============================================================================*/
#region Using directives

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Windows.Forms;
using Microsoft.SqlServer.Management.Common;
using Microsoft.SqlServer.Management.Smo;
using Microsoft.SqlServer.MessageBox;

#endregion

namespace Microsoft.Samples.SqlServer
{
    partial class DatabaseDefrag : Form
    {
        // Use the Server object to connect to a specific server
        private Server SqlServerSelection;

        public DatabaseDefrag()
        {
            InitializeComponent();
        }

        private void DatabaseDefrag_Load(object sender, EventArgs e)
        {
            ServerConnection ServerConn;
            ServerConnect scForm;
            DialogResult dr;

            Cursor csr = null;

            try
            {
                csr = this.Cursor;   // Save the old cursor
                this.Cursor = Cursors.WaitCursor;   // Display the waiting cursor

                // Display the main window first
                this.Show();
                Application.DoEvents();

                ServerConn = new ServerConnection();
                scForm = new ServerConnect(ServerConn);
                dr = scForm.ShowDialog(this);
                if ((dr == DialogResult.OK) &&
                    (ServerConn.SqlConnectionObject.State == ConnectionState.Open))
                {
                    SqlServerSelection = new Server(ServerConn);
                    if (SqlServerSelection != null)
                    {
                        this.Text = Properties.Resources.AppTitle
                            + SqlServerSelection.Name;

                        // Limit the table properties returned to just those that we use
                        SqlServerSelection.SetDefaultInitFields(typeof(Table),
                            new String[] { "Schema", "Name", "IsSystemObject" });

                        ShowDatabases(true);
                    }
                }
                else
                {
                    this.Close();
                }
            }
            catch (SmoException ex)
            {
                ExceptionMessageBox emb = new ExceptionMessageBox(ex);
                emb.Show(this);
            }
            finally
            {
                this.Cursor = csr;  // Restore the original cursor
            }
        }

        private void DatabaseDefrag_FormClosed(object sender, FormClosedEventArgs e)
        {
            if (SqlServerSelection != null)
            {
                if (SqlServerSelection.ConnectionContext.IsOpen == true)
                {
                    SqlServerSelection.ConnectionContext.Disconnect();
                }
            }
        }

        private void defragmentButton_Click(object sender, EventArgs e)
        {
            // Defragment the selected database
            DefragDatabase();
        }

        private void DefragDatabase()
        {
            Database db;
            Cursor csr = null;

            try
            {
                csr = this.Cursor;   // Save the old cursor
                this.Cursor = Cursors.WaitCursor;   // Display the waiting cursor

                // Clear control
                richTextBox1.Clear();

                db = (Database)DatabasesComboBox.SelectedItem;
                foreach (Table tbl in db.Tables)
                {
                    if (tbl.IsSystemObject == false)
                    {
                        richTextBox1.AppendText(Properties.Resources.Table
                            + tbl.ToString() + Environment.NewLine);

                        richTextBox1.ScrollToCaret();

                        // Allow the text to display
                        Application.DoEvents();

                        tbl.RebuildIndexes(90);
                    }
                }
                //  At last display the "completed"
                richTextBox1.AppendText(Properties.Resources.Completed);
            }
            catch (SmoException ex)
            {
                ExceptionMessageBox emb = new ExceptionMessageBox(ex);
                emb.Show(this);
            }
            finally
            {
                this.Cursor = csr;  // Restore the original cursor
            }
        }

        private void ShowDatabases(bool selectDefault)
        {
            // Show the current databases on the server
            Cursor csr = null;

            try
            {
                csr = this.Cursor;   // Save the old cursor
                this.Cursor = Cursors.WaitCursor;   // Display the waiting cursor

                // Clear control
                DatabasesComboBox.Items.Clear();

                // Limit the database properties returned to just those that we use
                SqlServerSelection.SetDefaultInitFields(typeof(Database),
                    new String[] { "Name", "IsSystemObject", "IsAccessible" });

                // Add database objects to combobox; the default ToString will display the database name
                foreach (Database db in SqlServerSelection.Databases)
                {
                    if (db.IsSystemObject == false && db.IsAccessible == true)
                    {
                        DatabasesComboBox.Items.Add(db);
                    }
                }

                if ((selectDefault == true) &&
                    (DatabasesComboBox.Items.Count > 0))
                {
                    DatabasesComboBox.SelectedIndex = 0;
                }
            }
            catch (SmoException ex)
            {
                ExceptionMessageBox emb = new ExceptionMessageBox(ex);
                emb.Show(this);
            }
            finally
            {
                this.Cursor = csr;  // Restore the original cursor
            }
        }
    }
}

File: SMO/DatabaseDefrag/VB/DatabaseDefrag/DatabaseDefrag.vb
'============================================================================
'  File:    DatabaseDefrag.vb 
'
'  Summary: Implements a sample SMO SQL Server database defragment utility in VB.NET.
'
'  Date:    August 29, 2005
'------------------------------------------------------------------------------
'  This file is part of the Microsoft SQL Server Code Samples.
'
'  Copyright (C) Microsoft Corporation.  All rights reserved.
'
'  This source code is intended only as a supplement to Microsoft
'  Development Tools and/or on-line documentation.  See these other
'  materials for detailed information regarding Microsoft code samples.
'
'  THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY
'  KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
'  IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
'  PARTICULAR PURPOSE.
'============================================================================
#Region "Using directives"

Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Windows.Forms
Imports Microsoft.SqlServer.Management.Common
Imports Microsoft.SqlServer.Management.Smo
Imports Microsoft.SqlServer.MessageBox


#End Region

Partial Class DatabaseDefrag 'The Partial modifier is only required on one class definition per project.
    Inherits Form
    ' Use the Server object to connect to a specific server
    Private SqlServerSelection As Server

    Public Sub New()
        InitializeComponent()
    End Sub

    Private Sub DatabaseDefrag_Load(ByVal sender As Object, ByVal e As EventArgs) Handles Me.Load
        Dim ServerConn As ServerConnection
        Dim scForm As ServerConnect
        Dim dr As DialogResult

        Dim csr As Cursor = Nothing

        Try
            csr = Me.Cursor ' Save the old cursor
            Me.Cursor = Cursors.WaitCursor ' Display the waiting cursor
            ' Display the main window first
            Me.Show()
            Application.DoEvents()

            ServerConn = New ServerConnection()
            scForm = New ServerConnect(ServerConn)
            dr = scForm.ShowDialog(Me)
            If dr = Windows.Forms.DialogResult.OK _
                AndAlso ServerConn.SqlConnectionObject.State = ConnectionState.Open Then
                SqlServerSelection = New Server(ServerConn)
                If Not (SqlServerSelection Is Nothing) Then
                    Me.Text = My.Resources.AppTitle & SqlServerSelection.Name

                    ' Limit the table properties returned to just those that we use
                    SqlServerSelection.SetDefaultInitFields(GetType(Table), _
                        New String() {"Schema", "Name", "IsSystemObject"})

                    ShowDatabases(True)
                End If
            Else
                Me.Close()
            End If
        Catch ex As SmoException
            Dim emb As New ExceptionMessageBox(ex)
            emb.Show(Me)
        Finally
            Me.Cursor = csr ' Restore the original cursor
        End Try
    End Sub

    Private Sub DatabaseDefrag_FormClosed(ByVal sender As Object, ByVal e As FormClosedEventArgs) Handles Me.FormClosed
        If Not (SqlServerSelection Is Nothing) Then
            If SqlServerSelection.ConnectionContext.IsOpen = True Then
                SqlServerSelection.ConnectionContext.Disconnect()
            End If
        End If
    End Sub

    Private Sub defragmentButton_Click(ByVal sender As Object, ByVal e As EventArgs) Handles defragmentButton.Click
        ' Defragment the selected database
        DefragDatabase()
    End Sub

    Private Sub DefragDatabase()
        Dim db As Database
        Dim csr As Cursor = Nothing

        Try
            csr = Me.Cursor ' Save the old cursor
            Me.Cursor = Cursors.WaitCursor ' Display the waiting cursor
            ' Clear control
            richTextBox1.Clear()

            db = CType(DatabasesComboBox.SelectedItem, Database)
            Dim tbl As Table
            For Each tbl In db.Tables
                If tbl.IsSystemObject = False Then
                    richTextBox1.AppendText(My.Resources.Table & tbl.ToString() & Environment.NewLine)

                    richTextBox1.ScrollToCaret()

                    ' Allow the text to display
                    Application.DoEvents()

                    tbl.RebuildIndexes(90)
                End If
            Next tbl
            '  At last display the 'Completed'
            richTextBox1.AppendText(My.Resources.Completed)
        Catch ex As SmoException
            Dim emb As New ExceptionMessageBox(ex)
            emb.Show(Me)
        Finally
            Me.Cursor = csr ' Restore the original cursor
        End Try
    End Sub

    Private Sub ShowDatabases(ByVal selectDefault As Boolean)
        ' Show the current databases on the server
        Dim csr As Cursor = Nothing

        Try
            csr = Me.Cursor ' Save the old cursor
            Me.Cursor = Cursors.WaitCursor ' Display the waiting cursor
            ' Clear control
            DatabasesComboBox.Items.Clear()

            ' Limit the database properties returned to just those that we use
            SqlServerSelection.SetDefaultInitFields(GetType(Database), New String() {"Name", "IsSystemObject", "IsAccessible"})

            ' Add database objects to combobox; the default ToString will display the database name
            Dim db As Database
            For Each db In SqlServerSelection.Databases
                If db.IsSystemObject = False AndAlso db.IsAccessible = True Then
                    DatabasesComboBox.Items.Add(db)
                End If
            Next db

            If selectDefault = True AndAlso DatabasesComboBox.Items.Count > 0 Then
                DatabasesComboBox.SelectedIndex = 0
            End If
        Catch ex As SmoException
            Dim emb As New ExceptionMessageBox(ex)
            emb.Show(Me)
        Finally
            Me.Cursor = csr ' Restore the original cursor
        End Try
    End Sub
End Class